mmgen/models/diffusions/base_diffusion.py

```diff
@@ -0,0 +1,903 @@
+
+        if sample_model == 'ema':
+            assert self.use_ema
+            _model = self.denoising_ema
+        elif sample_model == 'ema/orig' and self.use_ema:
+            _model = self.denoising_ema
+        else:
+            _model = self.denoising
+
+        outputs = sample_fn(
+            _model,
+            noise=noise,
+            num_batches=num_batches,
+            label=label,
+            **kwargs)
+
+        if isinstance(outputs, dict) and 'noise_batch' in outputs:
+            noise = outputs['noise_batch']
+
+        if sample_model == 'ema/orig' and self.use_ema:
+            _model = self.denoising
+            outputs_ = sample_fn(
+                _model, noise=noise, num_batches=num_batches, **kwargs)
+
+            if isinstance(outputs_, dict):
+                outputs['fake_img'] = torch.cat(
+                    [outputs['fake_img'], outputs_['fake_img']], dim=0)
+            else:
+                outputs = torch.cat([outputs, outputs_], dim=0)
+
+        return outputs
+
+    @torch.no_grad()
+    def DDPM_sample(self,
+                    model,
+                    noise=None,
+                    num_batches=0,
+                    label=None,
+                    save_intermedia=False,
+                    timesteps_noise=None,
+                    **kwargs):
+        """DDPM sample from random noise.
+        Args:
+            model (torch.nn.Module): Denoising model used to sample images.
+            noise (torch.Tensor | callable | None): You can directly give a
+                batch of noise through a ``torch.Tensor`` or offer a callable
+                function to sample a batch of noise data. Otherwise, the
+                ``None`` indicates to use the default noise sampler.
+            num_batches (int, optional):
+            label (torch.Tensor, optional):
+            save_intermedia (bool, optional): Whether to save denoising result
+                of intermedia timesteps. If set as True, would return a list
+                contains denoising results of each timestep. Otherwise, only
+                the final denoising result would be returned. Defaults to
+                False.
+            timesteps_noise (torch.Tensor, optional): Noise term used in each
+                denoising timesteps. If passed, the input noise would shape as
+                [num_timesteps, b, c, h, w]. If set as None, noise of each
+                denoising timestep would be randomly sampled. Default as None.
+        Returns:
+            torch.Tensor | list[torch.Tensor]: If ``save_intermedia``, a list
+                contains denoising results of each timestep would be returned.
+                Otherwise, only the final denoising result would be returned.
+        """
+        device = get_module_device(self)
+        x_t = self.get_noise(noise, num_batches=num_batches).to(device)
+        if save_intermedia:
+            # save input
+            intermedia = [x_t.clone()]
```

Comment:
In general, we will use `dict` to contain the intermediate results with the key indicating information for the features.